src/defrag.c

```diff
@@ -1016,30 +1021,39 @@ void activeDefragCycle(void) {
 
             db = &server.db[current_db];
             cursor = 0;
+            expires_cursor = 0;
             slot = findSlotByKeyIndex(db, 1, DB_MAIN);
+            defrag_later_item_in_progress = 0;
             ctx.db = db;
             ctx.slot = slot;
         }
         do {
             dict *d = db->dict[slot];
             /* before scanning the next bucket, see if we have big keys left from the previous bucket to scan */
-            if (defragLaterStep(db, endtime)) {
+            if (defragLaterStep(db, slot, endtime)) {
                 quit = 1; /* time is up, we didn't finish all the work */
                 break; /* this will exit the function and we'll continue on the next cycle */
             }
 
-            /* Scan the keyspace dict unless we're scanning the expire dict. */
-            if (!expires_cursor)
-                cursor = dictScanDefrag(d, cursor, defragScanCallback,
-                                        &defragfns, &ctx);
-            /* When done scanning the keyspace dict, we scan the expire dict. */
-            if (!cursor)
-                expires_cursor = dictScanDefrag(db->expires[slot], expires_cursor,
-                                                scanCallbackCountScanned,
-                                                &defragfns, NULL);
+            if (!defrag_later_item_in_progress) {
```

Comment:
defragLaterStep returns true if more work is needed.
so why do we need this?

taking a closer look, i think the code before this change already handled the defragLater after each dict.
am i missing something?

Reply:
In the earlier logic, the next slot discovery used to happen as soon as the main/expires dict scan was completed. And the defrag_later list would be non-empty at that point. Like we discussed above, we wanted to reuse the slot information in the defragLaterStep and avoid the recomputation. This change avoids the next slot discovery until the defrag_later list becomes empty.

Reply:
i don't see that in the code. one of us is missing something, so let's add more details.
i see two `do` loops, and the first thing that happens in both is a `defragLaterStep` and a `break` if it returns true.
the only thing we need to worry about is that we don't switch slot when that list isn't empty, and for that purpose, we just need to make sure that the code that changes `ctx.slot` isn't reachable when the list isn't empty.
or in other words, the `continue` you added seems enough and there's no need for that variable and if-else you also added.

what am i missing?

Reply:
If we don't have the `if (!defrag_later_item_in_progress) {` condition it would scan the same slot again as `cursor` and `expires_cursor` would be set to `0`, which we want to avoid.